sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala

```diff
@@ -0,0 +1,199 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import scala.collection.Map
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.analysis._
+import org.apache.spark.sql.types._
+
+object GetField {
+  /**
+   * Returns the resolved `GetField`. It will return one kind of concrete `GetField`,
+   * depend on the type of `child` and `fieldExpr`.
+   */
+  def apply(
+      child: Expression,
+      fieldExpr: Expression,
+      resolver: Resolver): GetField = {
+
+    (child.dataType, fieldExpr) match {
+      case (StructType(fields), Literal(fieldName, StringType)) =>
+        val ordinal = findField(fields, fieldName.toString, resolver)
+        SimpleStructGetField(child, fields(ordinal), ordinal)
+      case (ArrayType(StructType(fields), containsNull), Literal(fieldName, StringType)) =>
+        val ordinal = findField(fields, fieldName.toString, resolver)
+        ArrayStructGetField(child, fields(ordinal), ordinal, containsNull)
+      case (_: ArrayType, _) if fieldExpr.dataType.isInstanceOf[IntegralType]  =>
+        ArrayOrdinalGetField(child, fieldExpr)
+      case (_: MapType, _) =>
+        MapOrdinalGetField(child, fieldExpr)
+      case (otherType, _) =>
+        throw new AnalysisException(
+          "GetField is not valid on child of type " +
+            s"$otherType with fieldExpr of type ${fieldExpr.dataType}")
+    }
+  }
+
+  def unapply(g: GetField): Option[(Expression, Expression)] = {
+    g match {
+      case _: StructGetField => Some((g.child, null))
+      case o: OrdinalGetField => Some((o.child, o.ordinal))
+      case _ => None
+    }
+  }
+
+  /**
+   * find the ordinal of StructField, report error if no desired field or over one
+   * desired fields are found.
+   */
+  private def findField(fields: Array[StructField], fieldName: String, resolver: Resolver): Int = {
+    val checkField = (f: StructField) => resolver(f.name, fieldName)
+    val ordinal = fields.indexWhere(checkField)
+    if (ordinal == -1) {
+      throw new AnalysisException(
+        s"No such struct field $fieldName in ${fields.map(_.name).mkString(", ")}")
+    } else if (fields.indexWhere(checkField, ordinal + 1) != -1) {
+      throw new AnalysisException(
+        s"Ambiguous reference to fields ${fields.filter(checkField).mkString(", ")}")
+    } else {
+      ordinal
+    }
+  }
+}
+
+trait GetField extends UnaryExpression {
+  self: Product =>
+
+  type EvaluatedType = Any
+}
+
+abstract class StructGetField extends GetField {
```

Comment:
maybe we can remove this? I don't see a lot of benefits of having a common parent type for these 2 classes. On the contrary it makes the type hierarchy fairly complicated.